### ISSUE TITLE
Extensible arrows break partial lines in LaTeX tables

### ISSUE THREAD

**Issue by rbeezer** [rbeezer/mathbook#59]
I am implementing horizontal partial lines across tables, of variable thickness.  The minimal working example from tex.se is the basis of the implementation.  Add in the extensible arrow package `extpfeil` and it silently breaks.  The horizontal lines just become their default thickness.  Have not decided what to do about this conflict (tables should win, I think), but a fix would be even better.  I cannot see how the conflict arises.

http://tex.stackexchange.com/questions/24549/horizontal-rule-with-adjustable-height-behaving-like-clinen-m

```latex
\documentclass{book}

% \usepackage{extpfeil}

\newlength{\Oldarrayrulewidth}
\newcommand{\Cline}[2]{%
  \noalign{\global\setlength{\Oldarrayrulewidth}{\arrayrulewidth}}%
  \noalign{\global\setlength{\arrayrulewidth}{#1}}\cline{#2}%
  \noalign{\global\setlength{\arrayrulewidth}{\Oldarrayrulewidth}}}
\begin{document}

\begin{tabular}{ccc}
  a & b & c \\\Cline{2pt}{2-3}
  d & e & f \\\Cline{5pt}{1-2}
\end{tabular}

\end{document}
```

**Comment by cmhughes** [comment]
This is an interesting issue! I don't have a complete answer, but I have some information...

After some digging around, I found that it appears to be the `calc` package that causes the conflict; the `extpfeil` package loads `mathtools` which loads `calc`. So, for example, the following MWE also breaks:

```
\documentclass{book}

\usepackage{calc}
\newlength{\Oldarrayrulewidth}
\newcommand{\Cline}[2]{%
  \noalign{\global\setlength{\Oldarrayrulewidth}{\arrayrulewidth}}%
  \noalign{\global\setlength{\arrayrulewidth}{#1}}\cline{#2}%
  \noalign{\global\setlength{\arrayrulewidth}{\Oldarrayrulewidth}}}
\begin{document}

\begin{tabular}{ccc}
  a & b & c \\\Cline{2pt}{2-3}
  d & e & f \\\Cline{3pt}{1-2}
\end{tabular}

\end{document}
```

If you'd like a `cline`-like command that can take a width, then the `booktabs` package provides `cmidrule[<width>]{column-column}`, so that the following MWE works as expected.

```
% arara: pdflatex
\documentclass{book}

\usepackage{extpfeil}
\usepackage{booktabs}
\newlength{\Oldarrayrulewidth}
\newcommand{\Cline}[2]{%
  \noalign{\global\setlength{\Oldarrayrulewidth}{\arrayrulewidth}}%
  \noalign{\global\setlength{\arrayrulewidth}{#1}}\cline{#2}%
  \noalign{\global\setlength{\arrayrulewidth}{\Oldarrayrulewidth}}}
\begin{document}

\begin{tabular}{ccc}
  a & b & c \\\cmidrule[2pt]{2-3}
  d & e & f \\\cmidrule[4pt]{1-2}
\end{tabular}

\end{document}
```
This may not be perfect for your intended implmentation--if you'd prefer not to load the `booktabs` package, then you might copy the definition of `cmidrule` from `booktabs.sty`, although it's quite a small package as it is, so copying these lines doesn't save much.

I appreciate this approach isn't ideal; I'll continue to try to dig around and find out what the problem is as it would be nice to understand *why* this issue prevails.

**Comment by rbeezer** [comment]
Thanks very much, Chris.  Odder still.  I'm surprised the `calc` package does not bleed through somewhere else.

The `booktabs` package is wonderful and I was using it initially.  The author is militantly anti-vertical-rule, and I am a convert, but too many people want them, and think they need them.  I'll use MBX documentation as a bully pulpit, but also give the people what they want (and things like Cayley tables really do need vertical lines).  Therefore `booktabs` totally breaks (literally) vertical rules, so I think that it not going to be an option.

`extpfeil` was a user request and MathJax has an extension, so I may just "temporarily" disable the package on the latex side, though I really do not want to do that.  And maybe I should post on stackexchange where I saw the definitions that are causing the problem and a solution might arise (and at least others would be warned).  

Thanks very much for pursuing this one.

**Comment by cmhughes** [comment]
I can completely understand the need for a versatile project such as MBX to
support vertical lines.

I'll continue to dig, but a question on the tex exchange may get some good
traction, especially as you have such a good MWE that demonstrates the
problem so clearly.

**Comment by cmhughes** [comment]
After some more digging, I found that the `calc` package redefines the `setlength` command. One possible (ugly?) fix is to store and then use the original definition of `setlength` as follows:

```
\documentclass{book}

\let\oldsetlength\setlength
\usepackage{extpfeil}

\newlength{\Oldarrayrulewidth}
\newcommand{\Cline}[2]{%
  \noalign{\global\oldsetlength{\Oldarrayrulewidth}{\arrayrulewidth}}%
  \noalign{\global\oldsetlength{\arrayrulewidth}{#1}}\cline{#2}%
  \noalign{\global\oldsetlength{\arrayrulewidth}{\Oldarrayrulewidth}}}
\begin{document}

\begin{tabular}{ccc}
  a & b & c \\\Cline{2pt}{2-3}
  d & e & f \\\Cline{5pt}{1-2}
\end{tabular}

\end{document}
```

**Comment by rbeezer** [comment]
Excellent!  Ugly indeed, but I understand the fix, which means it cannot be so 
bad.  Seems a reckless thing for the calc package to do.

I think there are five or so "arrow" macros in the extpfeil package, so I think 
I can condition on their appearance, and/or on the presence of tables, and only 
write in the fix when it is needed.

This is one of the things I like about the LaTeX output - trapping these silly 
conflicts and putting in the work-arounds.  I'll merge the new code for tables 
in the next couple days and will test this out first.

Thanks very much.

Rob

**Comment by rbeezer** [comment]
Fix is in at:  5579a650

Thanks again, Chris, I didn't want to leave that hanging.
